internal/addon/authentication/provider.go

```diff
@@ -62,131 +44,77 @@ func NewSecretsProvider(k8s client.Client, clusterName string, signal addon.Sign
-		if err != nil {
-			klog.Error(err, "failed to configure resource")
+	secretKeys := make(map[Target]SecretKey, len(targetSecretTypes))
+	for targetName, authType := range targetSecretTypes {
+		switch authType {
+		case SecretReference:
+			obj, err := sp.discoverSecretRef(ctx, targetSecretName[targetName])
+			if err != nil {
+				return nil, err
+			}
+			secretKeys[targetName] = SecretKey(client.ObjectKeyFromObject(obj))
 			continue
-		}
-
-		msg := fmt.Sprintf("Resource has been %s", op)
-		switch op {
-		case ctrlutil.OperationResultNone:
-			klog.Info(msg)
 		default:
-			klog.Info(msg)
+			return nil, kverrors.New("missing mutate implementation for authentication type", "type", authType)
 		}
 	}
 
-	err := sp.injectCA(ctx, targetAuthType, secretKeys)
-	if err != nil {
-		return nil, err
-	}
-
 	return secretKeys, nil
 }
 
-// FetchSecrets given a map of Target and SecretKey it will get the Secret from
-// the hub cluster and add an annotation to it with Target. The goal of the
-// annotation is to preseve the link betweeen Target and Secret.
-// Note: the secret is not updated on the cluster with the annotation
-func (sp *secretsProvider) FetchSecrets(ctx context.Context, targetsSecret map[Target]SecretKey, targetAnnotation string) ([]corev1.Secret, error) {
-	secrets := make([]corev1.Secret, 0, len(targetsSecret))
-	for target, key := range targetsSecret {
+// FetchSecrets transforms a map of Target/SecretKey to a map of Target/Secret
+func (sp *secretsProvider) FetchSecrets(ctx context.Context, targetsSecrets map[Target]SecretKey) (map[Target]corev1.Secret, error) {
+	secrets := make(map[Target]corev1.Secret, len(targetsSecrets))
+	for target, key := range targetsSecrets {
 		secret := &corev1.Secret{}
 		if err := sp.k8s.Get(ctx, client.ObjectKey(key), secret, &client.GetOptions{}); err != nil {
 			return secrets, err
 		}
-		if secret.Annotations == nil {
-			secret.Annotations = make(map[string]string)
-		}
-		secret.Annotations[targetAnnotation] = string(target)
-		secrets = append(secrets, *secret)
+		secrets[target] = *secret
 	}
 	return secrets, nil
 }
 
-// injectCA will for Target's that requested mTLS authentication inject in the secret
-// an "ca-bundle.crt" key containing the CA configured in the secretsProvider Config
-func (sp *secretsProvider) injectCA(ctx context.Context, targetAuthType map[Target]AuthenticationType, targetsSecret map[Target]SecretKey) error {
-	if sp.MTLSConfig.CAToInject == "" {
-		return nil
-	}
-
-	objects := []client.Object{}
-	for target, authType := range targetAuthType {
-		switch authType {
-		case MTLS:
-			secret := &corev1.Secret{}
-			key := client.ObjectKey(targetsSecret[target])
-			if err := sp.k8s.Get(ctx, key, secret, &client.GetOptions{}); err != nil {
-				return err
-			}
-			manifests.InjectCA(secret, sp.MTLSConfig.CAToInject)
-			objects = append(objects, secret)
+func (sp *secretsProvider) discoverSecretRef(ctx context.Context, secretName string) (*corev1.Secret, error) {
```

Comment:
I would change the name of this method to `getSecretReference` or, at least, add a docstring.